### ISSUE TITLE
MP-08: Network Testing, Latency Simulation, and Polish

### ISSUE THREAD

**Issue by tombackert** [tombackert/ninja-game#145]
## Overview
Comprehensive testing, latency simulation, and polish for the multiplayer implementation. This ensures the networking code handles real-world conditions like packet loss, latency, and jitter.

**Related Document:** `docs/multiplayer-brainstorm.md` - Section 5 (Phase 5)

---

### Files to Create
- `scripts/network/latency_simulator.py` - Artificial latency/packet loss injection
- `tests/integration/test_multiplayer_e2e.py` - End-to-end multiplayer tests
- `experiments/network_stress_test.py` - Performance benchmarks

### Files to Modify
- `scripts/network/udp_transport.py` - Add debug hooks for simulation
- `scripts/network/game_client.py` - Add reconnection logic
- `scripts/network/game_server.py` - Add client timeout handling

#### Latency Simulator
```python
# scripts/network/latency_simulator.py
import random
import time
from collections import deque
from dataclasses import dataclass
from typing import Any, Deque, Tuple

@dataclass
class DelayedPacket:
    data: Any
    address: Tuple[str, int]
    deliver_time: float

class LatencySimulator:
    def __init__(
        self,
        min_latency_ms: float = 0,
        max_latency_ms: float = 0,
        packet_loss_percent: float = 0,
        jitter_ms: float = 0,
    ):
        self.min_latency = min_latency_ms / 1000
        self.max_latency = max_latency_ms / 1000
        self.packet_loss = packet_loss_percent / 100
        self.jitter = jitter_ms / 1000
        self.outgoing_queue: Deque[DelayedPacket] = deque()
        
    def should_drop_packet(self) -> bool:
        return random.random() < self.packet_loss
        
    def get_delay(self) -> float:
        base = random.uniform(self.min_latency, self.max_latency)
        jitter = random.uniform(-self.jitter, self.jitter)
        return max(0, base + jitter)
```

#### Reconnection Logic
```python
# In GameClient
class GameClient:
    CONNECTION_TIMEOUT_MS = 5000
    RECONNECT_ATTEMPTS = 3
    
    def _check_connection_health(self):
        elapsed = time.time() - self.last_server_message_time
        if elapsed > self.CONNECTION_TIMEOUT_MS / 1000:
            self._handle_timeout()
```

---

### Latency Simulation
- [ ] `LatencySimulator` can inject configurable latency (0-500ms)
- [ ] `LatencySimulator` can simulate packet loss (0-50%)
- [ ] `LatencySimulator` can simulate jitter
- [ ] Simulation can be enabled/disabled at runtime

### Connection Resilience
- [ ] Client detects server timeout after 5 seconds
- [ ] Client attempts reconnection (3 attempts)
- [ ] Server removes inactive clients after timeout
- [ ] Graceful disconnect sends notification

### Performance
- [ ] Server handles 4 clients at 60 Hz without frame drops
- [ ] Client maintains 60 fps with 100ms simulated latency
- [ ] Snapshot delta compression reduces bandwidth by >50%

### Edge Cases
- [ ] Game handles player joining mid-game
- [ ] Game handles player leaving mid-game

---

## Definition of Done
- [ ] All acceptance criteria met
- [ ] Test suite passes (`pytest -q`)
- [ ] End-to-end integration tests pass
- [ ] Stress test benchmarks documented
- [ ] `docs/architecture.md` updated with resilience patterns

---

### TC-01: Latency Injection
```python
def test_latency_simulator_delays_packets():
    sim = LatencySimulator(min_latency_ms=50, max_latency_ms=50)
    sim.queue_outgoing("test", ("127.0.0.1", 7777))
    assert len(sim.get_ready_packets()) == 0
    time.sleep(0.06)
    assert len(sim.get_ready_packets()) == 1
```

### TC-02: Client Timeout Detection
```python
def test_client_detects_timeout():
    client = GameClient(...)
    client.last_server_message_time = time.time() - 10
    client._check_connection_health()
    assert client.connection_state == ConnectionState.RECONNECTING
```

### TC-03: End-to-End with Simulated Latency
```python
def test_gameplay_with_100ms_latency():
    simulator = LatencySimulator(min_latency_ms=100, max_latency_ms=100)
    # Run 600 frames, verify sync within tolerance
```

### TC-04: Player Join/Leave Mid-Game
```python
def test_player_join_mid_game():
    # Start with 1 player, add second, verify both see each other
    
def test_player_leave_mid_game():
    # Start with 2 players, disconnect one, verify cleanup
```

---

## Dependencies
- MP-01 through MP-07 (all previous multiplayer issues)

---

## Performance Benchmarks

| Metric | Target |
|--------|--------|
| Server tick time (4 players) | < 5ms |
| Client tick time | < 3ms |
| Snapshot size (delta) | < 500 bytes |
| Snapshot size (full) | < 2 KB |
| Max supported latency | 250ms |
| Max packet loss tolerance | 10% |
